src/ott/solvers/nn/neuraldual.py

```diff
@@ -543,6 +573,43 @@ def g_value_finetuned(y: jnp.ndarray) -> jnp.ndarray:
         corr=True
     )
 
+  def resample(
+      self, batch_source: jnp.ndarray, batch_target: jnp.ndarray,
+      rng: jnp.ndarray
+  ) -> Tuple[jnp.ndarray, jnp.ndarray]:
+    r"""Resample data based on posterior marginals of entropy-regularized optimal transport solution.
+
+    Args:
+      batch_source: Batch of the source distribution.
+      batch_target: Batch of the target distribution.
+      rng: :class:`jax.numpy.ndarray` used for setting the seed.
+
+    Returns:
+      Resampled `batch_source` and `batch_target`.
+    """
+    if self.tau_a == 1.0 and self.tau_b == 1.0:
+      return batch_source, batch_target
+    geom = PointCloud(
+        batch_source,
+        batch_target,
+        epsilon=self.epsilon,
+        scale_cost=self.sample_sinkhorn_kwargs.pop("scale_cost", "mean")
+    )
+    prob = LinearProblem(geom, tau_a=self.tau_a, tau_b=self.tau_b)
+    out = sinkhorn.Sinkhorn(**self.sample_sinkhorn_kwargs)(prob)
+
+    # jax categorical uses log probabilities
+    log_marginals_source = jnp.log(jnp.sum(out.matrix, axis=1))
```

Comment:
Use `jnp.log(out.marginals(1))`

Reply:
Sorry, I couldn't find `marginals` to be an attribute of the `SinkhornOutput`, could you direct me to that please?

Reply:
Sorry, there was a typo, it's `out.marginal(1)`, see [here](https://github.com/ott-jax/ott/blob/main/src/ott/solvers/linear/sinkhorn.py#L399).